### DISCUSSION TITLE
Negate depending on Image Brightness

### DISCUSSION THREAD

**Discussion by bowespublishing** [ImageMagick/ImageMagick#3396] [Help]
I have a batch of images and depending on if they are dark images or light-coloured images I want to -channel RGB -negate (on the dark images only)

Is this possible?

Thank you

**Comment by fmw42** [comment]
Check the mean (average) value of the image before your processing and then decide if too dark, skip processing. You can check the mean value using

`convert image -format "%[fx:mean>threshold?1:0]" info:`

if the result is 1, it is bright and if result is 0, it is dark. Your OS will determine how to do the test and check it as scripting is different on Windows from Linux/Mac OSX.

**Comment by snibgo** [comment]
What version of IM, on what platform?

You can do it in a simple command like this:
```
magick in.png ( +clone -negate ) -delete "%[fx:u.mean>0.5?1:0]" out.png
```
This creates a clone that is negated, so we then have two images. Then it deletes one of the images, depending on the mean of the first image. If that mean is more than 0,5 (it is light), the negated image is deleted. Otherwise (the first image is dark), the first image is deleted. The remaining single image is written to out.png.

Windows CMD syntax shown. Adjust for Windows BAT or bash.

**Reply by fmw42** [reply]
Neat for IM 7, but won't work in IM 6. So will depend upon IM version.

**Comment by GeeMack** [comment]
The solution offered by **snibgo** may be nearly as simple as it gets, but as mentioned it works on ImageMagick v7, but not v6. Here is a similar solution that should work on any version. This is in Windows CLI syntax...

    convert input.png -set option:switch "%[fx:(mean>0.5)*180]" ^
       ( +clone -negate -rotate 180 ) -append ^
       -distort SRT %[switch] -extent 100x50% +repage result.png

That reads in the input image and sets an IM variable named "switch" to 180 if the mean brightness is above 0.5, and to 0 otherwise.

Next inside the parentheses it clones the input image, negates it, and rotates it 180 degrees. After the parentheses that modified version is appended to the bottom of the input image. So you have the original image on top, and on the bottom is the rotated negated copy.

The "-distort" operation will rotate that image 180 degrees if the mean of the input image was greater than 0.5, otherwise it won't rotate it. That leaves either the input image or the modified copy as the top half depending on the value of that "switch" variable.

Finish by using "-extent" to keep just the top half, do a "+repage" to reset the page geometry, and write the output file.

For IMv7 use "magick" instead of "convert". To run it in a BAT script double the percent signs "%%".

To adapt for a *nix shell change the continued-line carets "^" to backslashes "\\", escape the parentheses with backslashes "\\(...\\)", and maybe quote or escape any other special characters.

**Comment by bowespublishing** [comment]
Thank you. I'm trying to work on making nonogram puzzles using ImageMagick so the needing to negate depending on brightness. As if the image is too dark the negate messes it up

magick image.png -channel RGB -negate -bordercolor white -border 1 -alpha set -channel alpha -fuzz 18%% -fill white -floodfill +0+0 white -trim +dither -colors 2 -grayscale Rec601Luma -filter Lanczos -resize 20x20 -depth 1 -type truecolor +repage result.png
